### ISSUE TITLE
multivariate Bernoulli segmentation fault

### ISSUE THREAD

**Issue by VertexC** [jmschrei/pomegranate#471]
Hi,

I am trying to set up a multivariate model which has K state and each state has E features for  Bernoulli.

I get segmentation fault "Segmentation fault: 11" on multivariate Bernoulli.  Btw, if I didn't add any transition at the begining, is the value of transition matrix evenly distributed?
```python
def Bernoulli():
    # make up data
    # preset parameters
    E = 5  # number of essays
    G = 100  # length of genome
    K = 3  # number of segment types

    # initialize the matrix
    x_m = np.zeros((E, G),dtype=np.int8)
    # initialize():
    g = 0
    for g in range(G):
        for e in range(E):
            x_m[e, g] = bool(random.getrandbits(1)) 
    
    model = HiddenMarkovModel()
    s1 = State(IndependentComponentsDistribution([BernoulliDistribution(0.5) for e in range(E)]))
    s2 = State(IndependentComponentsDistribution([BernoulliDistribution(0.5) for e in range(E)]))
    s3 = State(IndependentComponentsDistribution([BernoulliDistribution(0.5) for e in range(E)]))
    print(s1)
    model.add_states(s1, s2, s3)
    model.add_transition(model.start, s1, 1.0)
    model.add_transition(s1, s1, 0.5)
    model.add_transition(s1, s2, 0.5)
    model.add_transition(s2, s2, 0.5)
    model.add_transition(s2, s3, 0.5)
    model.add_transition(s3, s3, 0.5)
    model.add_transition(s3, model.end, 0.1)
    model.bake()
    model.fit(x_m.T.copy(), algorithm='baum-welch')
```

**Comment by jmschrei** [comment]
Howdy

Looks like x_m is not properly formatted. Currently it's a sequence of observations of size (n, 5). pomegranate takes in a list of sequences because typically one has many sequences that they'd like to train their model on. If you only have one sequence, all you need to do is change your script to put x_m in a list, like so:

```python
model.fit([x_m.T.copy()], algorithm='baum-welch')
```

When I do that, and set verbose=True, I get the following output:

```
[1] Improvement: 22.9331154486	Time (s): 0.000221
[2] Improvement: 2.94293832117	Time (s): 0.0001631
[3] Improvement: 0.342696378044	Time (s): 0.000159
[4] Improvement: 0.30199358075	Time (s): 0.000154
[5] Improvement: 0.274065241238	Time (s): 0.0001769
[6] Improvement: 0.130348143061	Time (s): 0.000149
[7] Improvement: 0.0373955983494	Time (s): 0.0001471
[8] Improvement: 0.00852396543803	Time (s): 0.0001469
[9] Improvement: 0.00182065233298	Time (s): 0.000144
[10] Improvement: 0.000384785578717	Time (s): 0.000139
[11] Improvement: 8.14403086338e-05	Time (s): 0.0001831
[12] Improvement: 1.72811201082e-05	Time (s): 0.000138
[13] Improvement: 3.67337804619e-06	Time (s): 0.00014
[14] Improvement: 7.81607411682e-07	Time (s): 0.00015
[15] Improvement: 1.66393306245e-07	Time (s): 0.0001731
[16] Improvement: 3.54322082785e-08	Time (s): 0.0002
[17] Improvement: 7.54567963668e-09	Time (s): 0.0001719
[18] Improvement: 1.6075887288e-09	Time (s): 0.000159
[19] Improvement: 3.42197381542e-10	Time (s): 0.0002131
Total Training Improvement: 26.9733855023
Total Training Time (s): 0.0041
```

Let me know if that worked for you. One way or another it should've raised an error, so I'll look into adding one in.

**Comment by jmschrei** [comment]
Howdy

It looks like you’re initializing all of the Bernoulli distributions to the
same value and also setting the transition matrix to be uniform. If you try
to run EM, it is equally likely for any sample to take any path through the
HMM and so it won’t learn anything at all. You can see that the emissions
for each feature are the same for each component as well, and probably also
equal to the mean of the column. The standard way to initialize the
transition matrix of a model is to either (1) use random positive values
normalized to sum to 1 or (2) uniform values, and some standard ways for
initializing the emissions of a model are to (a) use random values, (b) use
the mean of the columns, or (c) use k-means to identify initial clusters
irrespective of position in the sequence. However, 2 and b can't be used
together because then each component in the model is identical.

Let me know if you have any other questions.

**Comment by VertexC** [comment]
Hi Jacob,

As you mentioned a list of observation sequences. Is the out-of-core training, if I have a list of sequence as [A,B,C], and do the training with 
```python
fit([A,B,C].copy(), algorithm='baum-welch',verbose=True, max_iterations=15, batch_size=100, batches_per_epoch=10, n_jobs=4) 
```
Is the batch_size and batches_per_epoch performed for each single sequence independently?

**Comment by jmschrei** [comment]
Howdy

pomegranate uses a method for training models that involves summarizing a batch of data into sufficient statistics using the `summarize` method, and then updating the internal parameters of the model using these sufficient statistics using the `from_summaries` method. These sufficient statistics are additive, meaning that if you summarize two batches separately and add together the sufficient statistics, you will get the same resulting model as if you summarized just one large batch. This naturally gives you a way to do out-of-core learning, since you can load up one batch from disk, summarize it, discard it, load up the next batch, etc... until you've seen your entire (potentially massive) data set, and then update only at the end. 

To answer your specific questions, `batch_size` is the number of sequences that are included in each batch and `batches_per_epoch` are the number of batches to be summarized before one updates. This isn't the number of symbols in a single sequence that are seen, but the number of full sequences. If you have only 3 sequences then `batch_size` must be smaller than 3 to do out-of-core learning.

If you'd like to implement your own out-of-core learning scheme you should use the `summarize` and `from_summaries` methods in any way you'd like. It might give you more flexibility than what I have built-in.

Let me know if you have any other questions.

**Comment by jmschrei** [comment]
Are you still encountering any of these issues, or should I close this?

**Comment by VertexC** [comment]
Oh, sorry for leaving it opening. Everything works so far:)
